_posts/2019-01-23-multifile-python-functions.md

```diff
@@ -0,0 +1,213 @@
+└── wordcloud
+    ├── __init__.py
+    ├── handler.py
+    └── requirements.txt
+```
+
+## The word count implementation with relative imports
+
+We are going to add two more files to our function:
+
+```sh
+touch wordcloud/stopwords
+touch wordcloud/wordcloud.py
+```
+
+The project now looks like
+
+```txt
+.
+├── stack.yml
+├── template
+│   ├── ...
+└── wordcloud
+    ├── __init__.py
+    ├── handler.py
+    ├── requirements.txt
+    ├── stopwords
+    └── wordcloud.py
+```
+
+The `stopwords` is a plain text file of words that will be excluded in the wordcount. These are short common words that you would not want to include in a wordcloud visualization such as: `a`, `an`, `him`, `her`, and so on. This list of words will depend on your use case, adjust as you see fit.
+
+All of the real fun happens in `wordcloud.py`:
+
+```py
+# modified wordcloud.py
+import unicodedata
+import os
+from typing import Dict, List
+
+from operator import itemgetter
+from collections import defaultdict
+
+FILE = os.path.dirname(__file__)
+STOPWORDS = set(
+    map(str.lower,
+        map(str.strip,
+            open(os.path.join(FILE, 'stopwords'), 'r').readlines()
+            )
+        )
+)
+
+
+def process_text(text: bytes) -> Dict[str, int]:
+    """Splits a long text into words a count of interesting words in the text.
+
+    process_text will eliminate any of the stopwords, punctuation, and normalize
+    the text to merge cases and plurals into a single value.
+    """
+    words = text.decode("utf-8").split()
+    # remove stopwords
+    # remove 's
+    words = [
+        word[:-2] if word.lower().endswith(("'s",))
+        else word
+        for word in words
+    ]
+    # remove numbers
+    words = [word for word in words if not word.isdigit()]
+    words = [strip_punctuation(word) for word in words]
+    words = [word for word in words if word.lower() not in STOPWORDS]
+
+    return process_tokens(words)
+
+def process_tokens(words: List[str]) -> Dict[str, int]:
+    """Normalize cases and remove plurals.
+    """
+    # ...
+
+def strip_punctuation(text: str) -> str:
+    # ...
+```
+
+Note that processing the `STOPWORDS` at the start of the file means it will be loaded into memory once the package is imported. In this case, once OpenFaaS deploys the function.
+
+Using an explicit relative import, we can very easily use the `process_text` method to create a very simple `handler.py`:
+
+```py
+# handler.py
+import json
+from .wordcloud import process_text
+
+
+def handle(req):
+    """handle a request to the function
+    Args:
+        req (str): request body
+    """
+
+    return json.dumps(process_text(req))
+```
+
+This style of relative imports will work for any file or sub-package you include inside of your function folder. Additionally, your IDE and linter will be able to resolve the imported code correctly!
+
+## Setup OpenFaaS
+
+OpenFaaS on Kubeneretes is my go-to deployment because I can use the built in k8s cluster in Docker-for-Mac.
```

Comment:
You could mention that you prefer Docker for Mac (now called Docker Desktop, on Mac) but please remove the cloned setup instructions since this will be hard to maintain over time and require continual edits.

Instead link to the K8s and Swarm deployment pages.

Reply:
done